File: Leetcode/Python3/215. Kth Largest Element in an Array/solution.py
# https://leetcode.com/problems/kth-largest-element-in-an-array/


import heapq


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        heap = []
        for n in nums:
            # Change n to negative to make the heap max_heap.
            heapq.heappush(heap, -n)

        for _ in range(k - 1):
            heapq.heappop(heap)

        return -heapq.heappop(heap)


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        heapq.heapify(nums)  # heapify convert list to heap

        for _ in range(len(nums) - k):
            heapq.heappop(nums)

        return heapq.heappop(nums)


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        # nlargest() automatically uses heapify()
        return heapq.nlargest(nums, k)


class Solution:  # Use sort
    def findKthLargest(self, nums: List[int], k: int) -> int:
        nums.sort()
        return nums[-k]

File: Leetcode/C++/215. Kth Largest Element in an Array/solution.cpp

#include <vector>
#include <queue>

using namespace std;

class Solution
{
public:
    int findKthLargest(vector<int> &nums, int k)
    {
        priority_queue<int> pq;
        for (int &num : nums)
        {
            pq.push(-num);
        }

        for (int i = 0; i < (nums.size() - k); i++)
        {
            pq.pop();
        }
        return -pq.top();
    }
};

class Solution
{
public:
    int findKthLargest(vector<int> &nums, int k)
    {
        sort(nums.rbegin(), nums.rend());
        return nums[k - 1];
    }
};